### DISCUSSION TITLE
Unclear how to use directives with Programmatic schema

### DISCUSSION THREAD

**Discussion by joel-wenzel** [graphql-java/graphql-java#2689] [Q&A] [answered]
Using programmatic schema generation I am able to add directives to the schema via:
``` java
GraphQLSchema.Builder builder = new GraphQLSchema.Builder();

GraphQLDirective filter = GraphQLDirective.newDirective()
                                                             .name("filter")
                                                             .validLocation(Introspection.DirectiveLocation.FIELD)
                                                             .build();
builder.additionalDirective(filter)
builder.build();
```

Using a SchemaPrinter I can see it output in the generated sdl file. However it is not clear how or where to apply the implementation logic of the directive when using GraphQLSchema.Builder.

I see in the [documentation](https://www.graphql-java.com/documentation/sdl-directives) you can use ```SchemaDirectiveWiring``` but that seems to only be possible when using ```RuntimeWiring``` which, as far as I can tell, cant be used when using ```GraphQLSchema.Builder```.

Am I missing something?

**Comment by bbakerman** [comment, marked as answer]
SchemaDirectiveWiring is indeed only called during the "SDL" geenration process.

If you want to add logic for your `@filter` directive then you need to somehow change the "DataFetchers" that are attached to your programmatically built `GraphqlSchema`.

`graphql.schema.GraphQLCodeRegistry` is the place in the `GraphqlSchema` that holds the code for data fetchers.

You could use a vistor tool like `graphql.schema.SchemaTraverser` and visit every fields DataFecther and change them so they have "preamble" code that knows about your field level `@filter`.  Remember as a `Introspection.DirectiveLocation.FIELD` directive its a directive provided by the calling client - not one that is statically built on the server side.

Once you change data fetcher with your filter logic then you "transform" the `GraphQLCodeRegistry` to have the new code and then you transform the `GraphqlSchema` to have the new `GraphQLCodeRegistry`.

**Reply by joel-wenzel** [reply]
@bbakerman  Not the answer I was hoping for but thank you none the less. Rather than using the ```graphql.schema.SchemaTraverser``` would be simple enough to extend the default property data fetcher and set that as the new default data fetcher or would that be bad practice?

**Reply by bbakerman** [reply]
>  Rather than using the graphql.schema.SchemaTraverser would be simple enough to extend the default property data fetcher and set that as the new default data fetcher or would that be bad practice?

I dont think you can actually change the default code per se.  But GraphqlCodeRegistry had a `defaultDataFetcher` method in its build that allows you to replace the default with another one... which could of course delegate back to the default

```
/**
         * This is the default data fetcher factory that will be used for fields that do not have specific data fetchers attached.  By default
         * {@link graphql.schema.PropertyDataFetcher} is used but you can have your own default via this method.
         *
         * @param defaultDataFetcherFactory the default data fetcher factory used
         *
         * @return this builder
         */
        public Builder defaultDataFetcher(DataFetcherFactory<?> defaultDataFetcherFactory) {
  ```
